homeassistant/components/ecobee/__init__.py

```diff
@@ -1,123 +1,167 @@
+
+        self._hass = hass
+        self._entry = entry
+        self.ecobee = Ecobee(
+            config={ECOBEE_API_KEY: api_key, ECOBEE_REFRESH_TOKEN: refresh_token}
+        )
 
     @Throttle(MIN_TIME_BETWEEN_UPDATES)
     def update(self):
-        """Get the latest data from pyecobee."""
-        self.ecobee.update()
-        _LOGGER.debug("Ecobee data updated successfully")
-
-
-def setup(hass, config):
-    """Set up the Ecobee.
+        """Get the latest data from ecobee.com."""
+        from pyecobee import ExpiredTokenError
+
+        try:
+            self.ecobee.update()
+            _LOGGER.debug("Updating ecobee.")
+        except ExpiredTokenError:
+            _LOGGER.warn("ecobee update failed; attempting to refresh expired tokens.")
+            self.refresh()
+
+    def refresh(self) -> bool:
+        """Refresh ecobee tokens and update config entry."""
+        from pyecobee import ECOBEE_API_KEY, ECOBEE_REFRESH_TOKEN
+
+        _LOGGER.debug("Refreshing ecobee tokens and updating config entry.")
+        if self.ecobee.refresh_tokens():
+            self._hass.config_entries.async_update_entry(
+                self._entry,
+                data={
+                    CONF_API_KEY: self.ecobee.config[ECOBEE_API_KEY],
+                    CONF_REFRESH_TOKEN: self.ecobee.config[ECOBEE_REFRESH_TOKEN],
+                },
+            )
+            return True
+        else:
+            _LOGGER.error("Error updating ecobee tokens.")
+            return False
+
+
+async def async_populate_options(hass, config_entry):
+    """
+    Populate options for ecobee. Called by async_setup_entry.
 
-    Will automatically load thermostat and sensor components to support
-    devices discovered on the network.
+    Options will be initially set to the values specified in
+    configuration.yaml, if they exist, or, if not, the default.
     """
-    global NETWORK
+    try:
+        hold_temp = hass.data[DATA_ECOBEE_CONFIG].get(CONF_HOLD_TEMP, DEFAULT_HOLD_TEMP)
+    except KeyError:
+        hold_temp = DEFAULT_HOLD_TEMP
+
+    options = {CONF_HOLD_TEMP: hold_temp}
 
-    if "ecobee" in _CONFIGURING:
-        return
+    hass.config_entries.async_update_entry(config_entry, options=options)
 
-    # Create ecobee.conf if it doesn't exist
-    if not os.path.isfile(hass.config.path(ECOBEE_CONFIG_FILE)):
-        jsonconfig = {"API_KEY": config[DOMAIN].get(CONF_API_KEY)}
-        save_json(hass.config.path(ECOBEE_CONFIG_FILE), jsonconfig)
 
-    NETWORK = EcobeeData(hass.config.path(ECOBEE_CONFIG_FILE))
+async def async_unload_entry(hass, config_entry):
+    """Unload the config entry, but store the ecobee API key for later."""
+    hass.data.pop(DOMAIN)
+    hass.data[DATA_ECOBEE_CONFIG] = {CONF_API_KEY: config_entry.data[CONF_API_KEY]}
 
-    setup_ecobee(hass, NETWORK.ecobee, config)
+    for platform in ECOBEE_PLATFORMS:
+        await hass.config_entries.async_forward_entry_unload(config_entry, platform)
```

Comment:
Use `asyncio.gather` to check that all platforms were unloaded concurrently. Something like this:
```py
tasks = []
for platform in ECOBEE_PLATFORMS:
    tasks.append(
        hass.config_entries.async_forward_entry_unload(config_entry, platform)
    )

return all(await asyncio.gather(*tasks))
```